pulsar-common/src/main/java/org/apache/pulsar/common/util/AsyncTokenBucket.java

```diff
@@ -0,0 +1,615 @@
+        }
+
+        @Override
+        public long getCapacity() {
+            return capacityFactor == 1.0d ? getRate() : (long) (getRate() * capacityFactor);
+        }
+
+        @Override
+        public long getRate() {
+            return rateFunction.getAsLong();
+        }
+    }
+
+    /**
+     * A builder class for creating instances of {@link DynamicRateAsyncTokenBucket}.
+     */
+    public static class DynamicRateAsyncTokenBucketBuilder
+            extends AsyncTokenBucketBuilder<DynamicRateAsyncTokenBucketBuilder> {
+        protected LongSupplier rateFunction;
+        protected double capacityFactor = 1.0d;
+        protected double initialFillFactor = 1.0d;
+        protected LongSupplier ratePeriodNanosFunction;
+        protected double targetFillFactorAfterThrottling = 0.01d;
+
+        protected DynamicRateAsyncTokenBucketBuilder() {
+        }
+
+        public DynamicRateAsyncTokenBucketBuilder rateFunction(LongSupplier rateFunction) {
+            this.rateFunction = rateFunction;
+            return this;
+        }
+
+        public DynamicRateAsyncTokenBucketBuilder ratePeriodNanosFunction(LongSupplier ratePeriodNanosFunction) {
+            this.ratePeriodNanosFunction = ratePeriodNanosFunction;
+            return this;
+        }
+
+        public DynamicRateAsyncTokenBucketBuilder capacityFactor(double capacityFactor) {
+            this.capacityFactor = capacityFactor;
+            return this;
+        }
+
+        public DynamicRateAsyncTokenBucketBuilder initialFillFactor(double initialFillFactor) {
+            this.initialFillFactor = initialFillFactor;
+            return this;
+        }
+
+        public DynamicRateAsyncTokenBucketBuilder targetFillFactorAfterThrottling(
+                double targetFillFactorAfterThrottling) {
+            this.targetFillFactorAfterThrottling = targetFillFactorAfterThrottling;
+            return this;
+        }
+
+        @Override
+        public AsyncTokenBucket build() {
+            return new DynamicRateAsyncTokenBucket(this.capacityFactor, this.rateFunction,
+                    this.clockSource,
+                    this.ratePeriodNanosFunction, this.resolutionNanos,
+                    this.initialFillFactor,
+                    targetFillFactorAfterThrottling);
+        }
+    }
+
+    /**
+     * Interface for a clock source that returns a monotonic time in nanoseconds with a required precision.
+     */
+    public interface MonotonicClockSource {
+        /**
+         * Returns the current monotonic clock time in nanoseconds.
+         *
+         * @param highPrecision if true, the returned value must be a high precision monotonic time in nanoseconds.
+         *                      if false, the returned value can be a granular precision monotonic time in nanoseconds.
+         * @return the current monotonic clock time in nanoseconds
+         */
+        long getNanos(boolean highPrecision);
```

Comment:
+1, I'll rename it. `getCurrentTickNanos` perhaps. The reason for this is that `time` could be confusing since it's not the time of the day (like it is in `System.currentTimeMillis()`. I think `tick` might be a better description of what the returned value is.

Reply:
Renaming to `getTickNanos`

Reply:
Done.